File: leetcode-submissions/3394_Check_if_Grid_can_be_Cut_into_Sections.ts
function checkValidCuts(n: number, rectangles: number[][]): boolean {
    const x_axis :number[][]= [ ];
    const y_axis : number[][]= [];
    for(const [x1, y1 , x2, y2] of rectangles){
        x_axis.push([x1,x2]);
        y_axis.push([y1, y2]);
    }
    const merge_x:number[][] = merge(x_axis);
    const merge_y:number[][] = merge(y_axis);
    return merge_x.length>=3 || merge_y.length>=3;
};
function merge(intervals:number[][]){
    intervals.sort((a, b)=>a[0]-b[0]);
    const result:number[][]= [intervals[0]];
    for(let i:number=1;i<intervals.length;i++){
        const lastInterval = result[result.length-1];
        if(lastInterval[1]>intervals[i][0]){
            lastInterval[0] = Math.min(lastInterval[0], intervals[i][0]);
            lastInterval[1]= Math.max(lastInterval[1], intervals[i][1]);

        }else{
            result.push(intervals[i]);
        }
    }
    return result;
};

File: leetcode-submissions/3394_Check_if_Grid_can_be_Cut_into_Sections.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>>result;
        result.push_back(intervals[0]);
        for(int i=1;i<intervals.size();i++){
            if(result.back()[1]>intervals[i][0]){
                result.back()[0] = min(intervals[i][0], result.back()[0]);
                result.back()[1]= max(intervals[i][1], result.back()[1]);
            }else{
                result.push_back(intervals[i]);
            }
        }
        return result ;
    }
    bool checkValidCuts(int n, vector<vector<int>>& rectangles) {
        // we need to create two intervals array one for x axis and another for y axis and check whether the atleast three cuts can be made on any of them 
        vector<vector<int>>x_axis, y_axis;
        // [startx, starty, endx, endy]
        // [x1, y1, x2, y2]
        for(int i=0;i<rectangles.size();i++){
            int x1 = rectangles[i][0];
            int y1 = rectangles[i][1];
            int x2 = rectangles[i][2];
            int y2 = rectangles[i][3];
            x_axis.push_back({x1, x2});
            y_axis.push_back({y1 , y2});
        }
        vector<vector<int>>result1 = merge(x_axis);
        vector<vector<int>>result2 = merge(y_axis);
        return result1.size()>=3 or result2.size()>=3;
    }
};